sqlwhat/checks/check_funcs.py

```diff
@@ -417,4 +417,90 @@ def check_result(state):
     state1 = lowercase(state)
     state2 = check_all_columns(state1)
     has_equal_value(state2)
-    return state2
\ No newline at end of file
+    return state2
+
+def check_query(state, query, selector, result,
+                error_msg="Running `{{query}}` after your submission generated an error.",
+                incorrect_msg="Running `{{query}}` after your submission didn't give the expected result."):
+    """Run arbitrary queries against to the DB connection to verify the database state.
+
+    For queries that do not return any output (INSERTs, UPDATEs, ...),
+    you cannot use functions like ``check_col()`` and ``is_equal()`` to verify the query result.
+
+    ``check_query()`` will rerun the student query in the transaction prepared by sqlbackend,
+    and immediately afterwards run a query that you specify in ``query``.
+    The result of this query, an ordered dictionary with tuples representing columns as values,
+    will be mutated with a unary function you provide in ``selector``,
+    after which it is compared to the target result you specify in ``result``.
+
+    Args:
+        query: A SQL query as a string that is executed after the student query is re-executed.
+        selector: A unary function that takes an OrderedDict (keys are column names as strings,
+                  values are column contents as tuples) and zooms in on the piece of interest.
+        result: The value that the 'selected part' of the query should match with.
+        error_msg: if specified, this overrides the automatically generated feedback
+                   message in case the query generated an error.
+        incorrect_msg: if specified, this overrides the automatically generated feedback
+                       message in case the selected part of the query result doesn't match with result.
+
+    :Example:
+
+        Suppose we are checking whether an INSERT happened correctly: ::
+
+            INSERT INTO company VALUES (2, 'filip', 28, 'sql-lane', 42)
+
+        We can write the following SCT: ::
+
+            Ex().check_query(query = 'SELECT COUNT(*) AS c FROM company',
+                             selector=lambda x: x['c'][0],
```

Comment:
This is interesting, but the `selector` + `result` syntax is very different to other testwhat and pythonwhat. Would it be possible for me to write something like

```python
Ex().check_query(query = "SELECT * FROM company where name = 'filip AND age = 28 AND address = 'sql-lane' AND answer = 42").check_result().has_equal_value()
```

So that I just check that the query returns the same result in both the student database state and the solution database state?

Reply:
@richierocks Great question. 

Unfortunately, `sqlbackend` only has a single connection to the database. When an exercise initializes, it runs the solution query against the database in a transaction, fetches the query result, and then rolls back the transaction. Any queries the student does after that are executed against the same connection, also in a transaction, and rolled back after it completed.
That's why I decided to be more explicit: say what you want to test about the database, fetch it, and say what it should return, explicitly.

However, (and I just thought about this), this shouldn't be necessary, even though we only have one connection. By using transactions, we could have `check_query()`:

- Run the student's submission + the query passed in `query` in a transaction and fetch the result.
- Run the solution submission + the query passed in `query` in a transaction and fetch the result.

This is definitely more time consuming, but as long as the databases are not crazy, this should be doable, and allows for **much** easier comparison.

In that case, to make it consistent with `pythonwhat`, it would be something like:

```python
Ex().has_equal_value(query = 'SELECT COUNT(*) AS c FROM company')
```

or, in your example:

```python
Ex().has_equal_value(query = "SELECT * FROM company where name = 'filip AND age = 28 AND address = 'sql-lane' AND answer = 42")
```

WDYT?

Reply:
Yes! I really like this `Ex().has_equal_value(query = whatever)` syntax.